docs/framework/react/guides/custom-errors.md

```diff
@@ -0,0 +1,267 @@
+---
+id: custom-errors
+title: Custom Errors
+---
+
+
+
+
+## Return String Values from Forms
+
+By default, TanStack Form allows you to return string error messages from validators:
+
+```tsx
+<form.Field
+  name="username"
+  validators={{
+    onChange: ({ value }) =>
+      value.length < 3 ? "Username must be at least 3 characters" : undefined
+  }}
+/>
+```
+
+For form-level validation affecting multiple fields:
+
+```tsx
+const form = useForm({
+  defaultValues: {
+    username: "",
+    email: ""
+  },
+  validators: {
+    onChange: ({ value }) => {
+      return {
+        fields: {
+          username: value.username.length < 3 ? "Username too short" : undefined,
+          email: !value.email.includes("@") ? "Invalid email" : undefined
+        }
+      };
+    }
+  }
+});
+```
+
+String errors are the most common type and are easily displayed in your UI:
+
+```tsx
+{field.state.meta.errors.map((error, i) => (
+  <div key={i} className="error">{error}</div>
+))}
+```
+
+## Return Non-String Values from Forms
+
+TanStack Form supports returning non-string values as errors, providing flexibility for different validation scenarios:
```

Comment:
Worth mentioning that any `truthy` value is handled as an error. For example returning 0 is considered as valid